3scaleAdapter/pkg/threescale/metrics/prometheus_test.go

```diff
@@ -0,0 +1,97 @@
+
+        `
+	r := NewMetricsReporter(true, 8080)
+	r.ObserveSystemLatency("www.fake.com", "123", time.Second+time.Millisecond)
+	// TODO - Uncomment when https://github.com/prometheus/client_golang/issues/494 is resolved
+	//err := testutil.CollectAndCompare(systemLatency, strings.NewReader(expect), metricName)
+	//if err != nil {
+	//	t.Fatalf(err.Error())
+	//}
+}
+
+func TestObserveBackendLatency(t *testing.T) {
+	const metricName = "threescale_backend_latency"
+	const expect = `
+		# HELP threescale_backend_latency Request latency for requests to 3scale backend
+		# TYPE threescale_backend_latency histogram
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.01"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.02"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.03"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.05"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.08"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.1"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.15"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.2"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.3"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="0.5"} 0
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="1"} 1
+		threescale_backend_latency_bucket{backendURL="www.fake.com",serviceID="123",le="+Inf"} 1
+		threescale_backend_latency_sum{backendURL="www.fake.com",serviceID="123"} 1
+		threescale_backend_latency_count{backendURL="www.fake.com",serviceID="123"} 1
+
+	`
+	r := NewMetricsReporter(true, 8080)
+	r.ObserveBackendLatency("www.fake.com", "123", time.Second)
+	// TODO - Uncomment when https://github.com/prometheus/client_golang/issues/494 is resolved
+	//err := testutil.CollectAndCompare(backendLatency, strings.NewReader(expect), metricName)
+	//if err != nil {
+	//	t.Fatalf(err.Error())
+	//}
+}
+
+func TestIncrementTotalRequests(t *testing.T) {
+	collector := totalRequests
+	if testutil.ToFloat64(collector) != 0 {
+		t.Fatalf("unexpected counter value for %s", collector.Desc().String())
+	}
+	r := NewMetricsReporter(true, 8080)
+	r.IncrementTotalRequests()
+	if testutil.ToFloat64(collector) != 1 {
+		t.Fatalf("unexpected counter value for %s", collector.Desc().String())
+	}
+}
+
+func TestIncrementCacheHits(t *testing.T) {
+	collector := cacheHits
+	if testutil.ToFloat64(collector) != 0 {
+		t.Fatalf("unexpected counter value for %s", collector.Desc().String())
+	}
+	r := NewMetricsReporter(true, 8080)
+	r.IncrementCacheHits()
+	if testutil.ToFloat64(collector) != 1 {
+		t.Fatalf("unexpected counter value for %s", collector.Desc().String())
+	}
+}
```

Comment:
can we also make sure incrementing is not just setting the value to 1? ie. increment several (random?) times and assert its value (also good to extract out the common code and pass in the collector)

Reply:
Done